internal/engine/build_and_deployer_test.go

```diff
@@ -793,6 +810,167 @@ func TestLiveUpdateWithRunFailureReturnsContainerIDs(t *testing.T) {
+	f.assertK8sUpsertCalled(true)
+
+	// should NOT have run live update
+	assert.Equal(t, 0, f.sCli.UpdateContainerCount)
+}
+
+func TestLiveUpdateMultipleImagesOneRunErrorExecutesRestOfLiveUpdatesAndDoesntImageBuild(t *testing.T) {
+	f := newBDFixture(t, k8s.EnvDockerDesktop, container.RuntimeDocker)
+	defer f.TearDown()
+
+	// First LiveUpdate will simulate a failed Run step
+	f.docker.ExecErrorsToThrow = []error{userFailureErr}
+
+	manifest, bs := multiImageLiveUpdateManifestAndBuildState(f)
+	result, err := f.bd.BuildAndDeploy(f.ctx, f.st, buildTargets(manifest), bs)
+	require.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Run step \"go install github.com/windmilleng/sancho\" failed with exit code: 123")
+
+	// one for each container update
+	assert.Equal(t, 2, f.docker.CopyCount)
+	assert.Equal(t, 2, len(f.docker.ExecCalls))
+
+	// expect NO image build
+	assert.Equal(t, 0, f.docker.BuildCount)
+	assert.Equal(t, 0, f.docker.PushCount)
+	f.assertK8sUpsertCalled(false)
+
+	// Make sure we returned the CIDs we LiveUpdated --
+	// they contain state now, we'll want to track them
+	liveUpdatedCIDs := result.LiveUpdatedContainerIDs()
+	expectedCIDs := []container.ID{"sancho-c", "sidecar-c"}
+	assert.ElementsMatch(t, expectedCIDs, liveUpdatedCIDs)
+}
+
+func TestLiveUpdateMultipleImagesOneUpdateErrorFallsBackToImageBuild(t *testing.T) {
+	f := newBDFixture(t, k8s.EnvDockerDesktop, container.RuntimeDocker)
+	defer f.TearDown()
+
+	// Second LiveUpdate will throw an error
+	f.docker.ExecErrorsToThrow = []error{nil, fmt.Errorf("whelp ¯\\_(ツ)_/¯")}
+
+	manifest, bs := multiImageLiveUpdateManifestAndBuildState(f)
+	_, err := f.bd.BuildAndDeploy(f.ctx, f.st, buildTargets(manifest), bs)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// one for each container update
+	assert.Equal(t, 2, f.docker.CopyCount)
+	assert.Equal(t, 2, len(f.docker.ExecCalls)) // second one errors
+
+	// expect image build (2x images) when we fall back from failed LiveUpdate
+	assert.Equal(t, 2, f.docker.BuildCount)
+	assert.Equal(t, 0, f.docker.PushCount)
+	f.assertK8sUpsertCalled(true)
+}
+
+func TestLiveUpdateMultipleImagesOneWithUnsyncedChangeFileFallsBackToImageBuild(t *testing.T) {
+	f := newBDFixture(t, k8s.EnvGKE, container.RuntimeDocker)
+	defer f.TearDown()
+
+	manifest, bs := multiImageLiveUpdateManifestAndBuildState(f)
+	bs[manifest.ImageTargetAt(1).ID()].FilesChangedSet["/not/synced"] = true // changed file not in a sync --> fall back to image build
+	_, err := f.bd.BuildAndDeploy(f.ctx, f.st, buildTargets(manifest), bs)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// should NOT have run live update
+	assert.Equal(t, 0, f.sCli.UpdateContainerCount)
+
+	// expect image build (2x images) when we fall back from failed LiveUpdate
+	assert.Equal(t, 2, f.docker.BuildCount)
+	assert.Equal(t, 2, f.docker.PushCount)
+	f.assertK8sUpsertCalled(true)
+}
+
+func multiImageLiveUpdateManifestAndBuildState(f *bdFixture) (model.Manifest, store.BuildStateSet) {
```

Comment:
seems weird to couple these into one function

Reply:
I mean, it's just giving me a specific handful of (pretty interconnected) test data. seems less weird than putting it in two different functions but still having the data it returns be so interconnected that the build state only makes sense for the specific manifest (cuz then the functions risk being called separately?) but also 🤷‍♀